crypto/ec/ec2_mult.c

```diff
@@ -315,6 +315,34 @@ static int ec_GF2m_montgomery_point_multiply(const EC_GROUP *group,
     if (!BN_GF2m_add(x2, x2, &group->b))
         goto err;               /* x2 = x^4 + b */
 
+    /* blinding: make sure z1 and z2 are independently blinded. */
+    do {
+        if (!BN_rand(z1, BN_num_bits(&group->field), 0, 0)) {
+            ECerr(EC_F_EC_GF2M_MONTGOMERY_POINT_MULTIPLY, ERR_R_BN_LIB);
+            return 0;
+        }
+    } while (BN_is_zero(z1));
+
+    /* first blind (x2,z2) using z1 as the random field element. */
+    if ((group->meth->field_encode != NULL
+         && !group->meth->field_encode(group, z1, z1, ctx))
+        || !group->meth->field_mul(group, x2, x2, z1, ctx)
+        || !group->meth->field_mul(group, z2, z2, z1, ctx))
+        return 0;
+
+    /* now generate another random field element to blind (x1,z1) */
+    do {
+        if (!BN_rand(z1, BN_num_bits(&group->field), 0, 0)) {
```

Comment:
I don't think this is correct. E.g. for B-233 the field polynomial will have 234 bits while field elements have 233 bits. I think it's

`BN_rand(&r->Z, BN_num_bits(&group->field) - 1, -1, 0))`

where I think the `top=-1` allows the top bit to be random and not fixed.

Can you verify in the debugger?

Reply:
Uhm, good point, let me check!

Reply:
FWIW this is what we're doing in 111 and master

https://github.com/openssl/openssl/blob/3cb55fe47c3398b81956e4fe20c4004524d47519/crypto/ec/ec2_smpl.c#L731-L738

Reply:
`BN_num_bits(group->field) - 1` makes total sense. We wanted to fix the top bit to guarantee that z1 and z2 had exactly the same length, but this is indeed not necessary anymore. I can commit and push both fixes shortly.